File: string-search/strategies/NaiveSearch.js
import SearchStrategy from "./SearchStrategy.js";

export default class NaiveSearch extends SearchStrategy {
  search(text, pattern, stepMode = false) {
    const steps = [];
    const matches = [];
    let comparisons = 0;

    if (!pattern.length || pattern.length > text.length) {
      return {
        algorithm: "Naive",
        matches,
        comparisons,
        textLength: text.length,
        patternLength: pattern.length,
        complexity: "O(n * m)",
        steps
      };
    }

    for (let i = 0; i <= text.length - pattern.length; i++) {
      let j = 0;

      if (stepMode) {
        steps.push(`Iniciando comparação na posição ${i}`);
      }

      while (j < pattern.length) {
        comparisons++;

        if (stepMode) {
          steps.push(`Comparando text[${i + j}] = "${text[i + j]}" com pattern[${j}] = "${pattern[j]}"`);
        }

        if (text[i + j] !== pattern[j]) {
          if (stepMode) {
            steps.push(`Falha na posição ${i + j}. Movendo padrão para ${i + 1}`);
          }
          break;
        }

        j++;
      }

      if (j === pattern.length) {
        matches.push(i);
        if (stepMode) {
          steps.push(`Padrão encontrado na posição ${i}`);
        }
      }
    }

    return {
      algorithm: "Naive",
      matches,
      comparisons,
      textLength: text.length,
      patternLength: pattern.length,
      complexity: "O(n * m)",
      steps
    };
  }
}

File: AlgoritmosDeBusca_String/app/strategies/naive_search.py
import time

from app.domain.search_result import SearchResult
from app.domain.search_strategy import SearchStrategy
from app.strategies._step_helper import add_step


class NaiveSearch(SearchStrategy):
    @property
    def name(self) -> str:
        return "Naive"

    @property
    def theoretical_complexity(self) -> str:
        return "O(n * m)"

    def search(
        self,
        text: str,
        pattern: str,
        include_steps: bool = False,
        max_steps: int = 300,
    ) -> SearchResult:
        start = time.perf_counter_ns()
        occurrences = []
        comparisons = 0
        steps = []

        n = len(text)
        m = len(pattern)

        if m == 0 or n == 0 or m > n:
            elapsed_ms = (time.perf_counter_ns() - start) / 1_000_000
            return SearchResult(
                algorithm=self.name,
                occurrences=[],
                comparisons=0,
                execution_time_ms=elapsed_ms,
                text_size=n,
                pattern_size=m,
                theoretical_complexity=self.theoretical_complexity,
                auxiliary_data={"observation": "Texto vazio, padrão vazio ou padrão maior que o texto."},
                steps=steps,
            )

        for i in range(n - m + 1):
            add_step(steps, f"Alinhando padrão na posição {i} do texto.", include_steps, max_steps)
            match = True

            for j in range(m):
                comparisons += 1
                add_step(
                    steps,
                    f"Comparando texto[{i + j}]='{text[i + j]}' com padrão[{j}]='{pattern[j]}'.",
                    include_steps,
                    max_steps,
                )

                if text[i + j] != pattern[j]:
                    match = False
                    add_step(steps, "Falha na comparação. Deslocando uma posição.", include_steps, max_steps)
                    break

            if match:
                occurrences.append(i)
                add_step(steps, f"Ocorrência encontrada na posição {i}.", include_steps, max_steps)

        elapsed_ms = (time.perf_counter_ns() - start) / 1_000_000
        return SearchResult(
            algorithm=self.name,
            occurrences=occurrences,
            comparisons=comparisons,
            execution_time_ms=elapsed_ms,
            text_size=n,
            pattern_size=m,
            theoretical_complexity=self.theoretical_complexity,
            auxiliary_data={"strategy": "Força bruta com deslocamento de uma posição por tentativa."},
            steps=steps,
        )
